toolkit/tools/imagegen/diskutils/initramfs.go

```diff
@@ -0,0 +1,229 @@
+func (i *InitramfsMount) Close() (err error) {
+	var bytesIO int
+
+	logger.Log.Debugf("Closing initramfs file '%s'", i.initramfsOutputFile.Name())
+
+	// Defer close calls to make sure we handle any errors, failing to
+	// close the file means we can't close the install root.
+	defer i.initramfsOutputFile.Close()
+	defer i.pgzWriter.Close()
+	defer i.cpioWriter.Close()
+
+	err = i.cpioWriter.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+	err = i.pgzWriter.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+
+	logger.Log.Debugf("Writing %d bytes to file", i.outputBuffer.Len())
+	bytesIO, err = i.initramfsOutputFile.Write(i.outputBuffer.Bytes())
+	if err != nil {
+		logger.Log.Errorf("Failed to write initramfs file: '%s'", err.Error())
+		return
+	}
+	logger.Log.Infof("Bytes writen to file: %d", bytesIO)
+
+	// Explicit call to fsync, archive corruption was occuring occasionally otherwise.
+	err = i.initramfsOutputFile.Sync()
+	if err != nil {
+		logger.Log.Errorf("Failed to sync initramfs file: '%s'", err.Error())
+		return
+	}
+
+	err = i.initramfsOutputFile.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+	return
+}
+
+// AddFileToInitramfs places a single file in the initramfs at the destination path.
+// - sourcePath: Path to file which is to be adde
+// - destPath: Final destination in the initramfs
+func (i *InitramfsMount) AddFileToInitramfs(sourcePath, destPath string) (err error) {
+	var bytesIO int64
+	fileInfo, err := os.Lstat(sourcePath)
+	file, err := os.Open(sourcePath)
+	if err != nil {
+		return
+	}
+	defer file.Close()
+
+	// Symlinks need to be resolved to their target file to be added to the cpio archive.
+	var link string
+	if fileInfo.Mode()&os.ModeSymlink != 0 {
+		link, err = os.Readlink(sourcePath)
+		if err != nil {
+			return
+		}
+
+		logger.Log.Debugf("--> Adding link: (%s) -> (%s)", sourcePath, link)
+	}
+
+	// Convert the OS header into a CPIO header
+	header, err := cpio.FileInfoHeader(fileInfo, link)
+	if err != nil {
+		return
+	}
+	header.Name = destPath
+
+	err = i.cpioWriter.WriteHeader(header)
+	if err != nil {
+		return
+	}
+
+	// Special files (unix sockets, directories, symlinks, ...) need to be handled differently
+	// since a simple byte transfer of the file's content into the CPIO archive can't be achieved.
+	if !fileInfo.Mode().IsRegular() {
+		// For a symlink the reported size will be the size (in bytes) of the link's target.
+		// Write this data into the archive.
+		if fileInfo.Mode()&os.ModeSymlink != 0 {
+			_, err = i.cpioWriter.Write([]byte(link))
+		}
+
+		// For all other special files, they will be of size 0 and only contain the header in the archive.
+		logger.Log.Debugf("Added special file %s", header.Name)
+		return
+	}
+
+	bytesIO, err = io.Copy(i.cpioWriter, file)
+	if err != nil {
+		return
+	}
```

Comment:
Suggestions: I'd reformat this a bit to a `if fileInfo.Mode().IsRegular() {...} else {...}` version for a few reasons:
1. Returning earlier in the code through one branch of `if` is something we were able to avoid in most places (except, of courece, for the `if err != nil` bits), which makes what we're doing here a bit unexpected and easy to miss.
2. It seems we're now not checking the error for `i.cpioWriter.Write` and can claim we've added a special file, while in reality that didn't happen. Checking for the error one time after the `if-else` would solve that in one go and keep the code short.

I'd also move the debug log that we've added a special file in front of the write and change the phrasing to "Adding...". The log at the end of this function would tell us if we succeeded or not regardless of the type of the file.

Reply:
Agreed on refactoring the conditional (I borrowed this from the older code, so if we don't switch to a single library we might want to have a clean up of that code as well at some point).

In theory the err values don't need to be checked since we are returning right after, but it would be easy to add more code and not realize that. I'll add explicit checks.

As for the logging order, most special files are just a header which was already written. But I'll move the log to handle the symlink case (which is the most common...)


---
In reply to: [550336493](https://github.com/microsoft/CBL-Mariner/pull/481#discussion_r550336493) [](ancestors = 550336493)